mev_inspect/arbitrages.py

```diff
@@ -23,70 +23,124 @@ def get_arbitrages(swaps: List[Swap]) -> List[Arbitrage]:
 
 
 def _get_arbitrages_from_swaps(swaps: List[Swap]) -> List[Arbitrage]:
-    pool_addresses = {swap.pool_address for swap in swaps}
+    """
+    An arbitrage is defined as multiple swaps in a series that result in the initial token being returned.
 
-    all_arbitrages = []
-
-    for index, first_swap in enumerate(swaps):
-        other_swaps = swaps[:index] + swaps[index + 1 :]
+    Input swap list does not have to be ordered and can have excess swaps but cannot miss swaps in the path of arbitrage.
+    Ex: [WETH -> UNI, UNI -> DAI, DAI -> WETH]
+    Non-working Ex: [WETH -> UNI, DAI -> WETH]
 
-        if first_swap.from_address not in pool_addresses:
-            arbitrage = _get_arbitrage_starting_with_swap(first_swap, other_swaps)
+    Algorithm works by first finding possible start and end swap pairs (strategy defined in _get_all_start_end_swaps) then finding all possible routes between them.
+    """
 
-            if arbitrage is not None:
-                all_arbitrages.append(arbitrage)
+    all_arbitrages = []
 
-    return all_arbitrages
+    start_ends = _get_all_start_end_swaps(swaps)
+    if len(start_ends) == 0:
+        return []
 
+    filtered_start_ends = filter(
+        lambda start_end: _swaps_have_valid_order_wrt_trace_address(list(start_end)),
+        start_ends,
+    )
+    for start_end in filtered_start_ends:
+        start = start_end[0]
+        end = start_end[1]
 
-def _get_arbitrage_starting_with_swap(
-    start_swap: Swap,
-    other_swaps: List[Swap],
-) -> Optional[Arbitrage]:
-    swap_path = [start_swap]
-    current_swap: Swap = start_swap
+        remaining_swaps = [swap for swap in swaps if swap != start and swap != end]
```

Comment:
I think it won't end up mattering here, but are these equality checks by attribute value comparisons or by the underlying object id?

it might be ideal to just return these as part of the generation of the start end pairs similar to the original version `(start_swap, end_swap, other_swaps)`

Reply:
Good call–Adjusted to "is not" to compare by identity. Just have to be careful not to copy the swap objects objects in later updates to the logic.

I also changed the naming of that list to `potential_intermediate_swaps`. I think it's more readable to filter for the `potential_intermediate_swaps` afterwards rather than expanding the tuple and returning the additional set.

Reply:
Yeah agreed on being "careful not to copy" - that concern is why I suggest generating the list of other swaps alongside since then it can be done using indices instead of a filtering of the original list like here https://github.com/flashbots/mev-inspect-py/blob/main/mev_inspect/arbitrages.py#L31